aptos-move/aptos-gas/src/instr.rs

```diff
@@ -0,0 +1,330 @@
+    [move_to_unit, "move_to.unit", 100],
+    [move_to_generic_base, "move_to_generic.base", 100],
+    [move_to_generic_unit, "move_to_generic.unit", 100],
+    // vec
+    [vec_len, "vec_len", 1],
+    [vec_imm_borrow, "vec_imm_borrow", 1],
+    [vec_mut_borrow, "vec_mut_borrow", 1],
+    [vec_push_back, "vec_push_back", 1],
+    [vec_pop_back, "vec_pop_back", 1],
+    [vec_swap, "vec_swap", 1],
+    [vec_pack_base, "vec_pack.base", 1],
+    [vec_pack_unit, "vec_pack.unit", 1],
+    [vec_unpack_base, "vec_unpack.base", 1],
+    [vec_unpack_unit, "vec_unpack.unit", 1],
+);
+
+impl InstructionGasParameters {
+    pub(crate) fn instr_cost(&self, op: Opcodes) -> PartialVMResult<u64> {
+        use Opcodes::*;
+
+        Ok(match op {
+            NOP => self.nop,
+
+            ABORT => self.abort,
+            RET => self.ret,
+
+            BR_TRUE => self.br_true,
+            BR_FALSE => self.br_false,
+            BRANCH => self.branch,
+
+            POP => self.pop,
+            LD_U8 => self.ld_u8,
+            LD_U64 => self.ld_u64,
+            LD_U128 => self.ld_u128,
+            LD_TRUE => self.ld_true,
+            LD_FALSE => self.ld_false,
+
+            IMM_BORROW_LOC => self.imm_borrow_loc,
+            MUT_BORROW_LOC => self.mut_borrow_loc,
+            IMM_BORROW_FIELD => self.imm_borrow_field,
+            MUT_BORROW_FIELD => self.mut_borrow_field,
+            IMM_BORROW_FIELD_GENERIC => self.imm_borrow_field_generic,
+            MUT_BORROW_FIELD_GENERIC => self.mut_borrow_field_generic,
+            FREEZE_REF => self.freeze_ref,
+
+            CAST_U8 => self.cast_u8,
+            CAST_U64 => self.cast_u64,
+            CAST_U128 => self.cast_u128,
+
+            ADD => self.add,
+            SUB => self.sub,
+            MUL => self.mul,
+            MOD => self.mod_,
+            DIV => self.div,
+
+            BIT_OR => self.bit_or,
+            BIT_AND => self.bit_and,
+            XOR => self.xor,
+            SHL => self.shl,
+            SHR => self.shr,
+
+            OR => self.or,
+            AND => self.and,
+            NOT => self.not,
+
+            LT => self.lt,
+            GT => self.gt,
+            LE => self.le,
+            GE => self.ge,
+
+            VEC_LEN => self.vec_len,
+            VEC_IMM_BORROW => self.vec_imm_borrow,
+            VEC_MUT_BORROW => self.vec_mut_borrow,
+            VEC_POP_BACK => self.vec_pop_back,
+            VEC_SWAP => self.vec_swap,
+
+            op => {
+                return Err(
+                    PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                        .with_message(format!("cannot charge gas for unknown operation {:?}", op)),
+                )
+            }
+        })
+    }
+
+    pub(crate) fn instr_cost_with_size(&self, op: Opcodes, size: u64) -> PartialVMResult<u64> {
+        use Opcodes::*;
+
+        Ok(match op {
+            LD_CONST => self.ld_const_base + self.ld_const_unit * size,
```

Comment:
All these entries look a bit redundant. Can we refactor this into a generic calculation function: base + unit * size?

Reply:
Probably not worth it. The best you can do right now is to have
```
let (base_cost, unit_cost) = match op {
    COPY_LOC => (self.copy_loc_base, self.copy_loc_unit),
    ...
};
let cost = base_cost + unit_cost * unit_cost
```
which is as complicated as what I already have.

Plus, we'll want to accommodate for formulae that don't fit into this model. Due to some legacy reasons, `VEC_PUSH_BACK` is treated as a sized instruction, but its cost model should only consist a base cost. We'll likely want to refine some the formulae a bit more in the future, meaning that more gas parameters may be added.